File: ruby/table.rb
class Table
  def initialize(hash)
    @hash = hash
    @mutex = Mutex.new
  end

  def put(key, value)
    @mutex.synchronize do
      @hash[key] = value 
    end
  end

  def get(key)
     @mutex.synchronize do
       @hash[key]
     end
  end 

  def delete(key)
    @mutex.synchroinze do
      @hash.delete key
    end
  end
end


def concurrent_test(count)
  table = Table.new({counter: 0})
  (1..count).map do |it|
    if it.even?
      Thread.new do
        table.put(:counter, 1)
      end
    else
      Thread.new do
        v = table.get(:counter)
        table.put(:counter, 2)
      end
    end
  end.each(&:join)
  table
end



def concurrent_test2(count)
  table = Table.new({counter: 0})
  (1..count).map do |it|
    if it.even?
      Thread.new do
        v = table.get(:counter)
        table.put(:counter, v+1)
      end
    else
      Thread.new do
        v = table.get(:counter)
        table.put(:counter, v-1)
      end
    end
  end.each(&:join)
  table
end

File: elixir/table.ex
require Integer

defmodule Table do
  def start(map) do
    spawn fn -> loop(map) end
  end

  def loop(map) do
    receive do

      {:put, key, value} ->
          loop(Map.put(map, key, value))

      {:get, key, ref} ->
          send ref, {:ok, Map.get(map, key)}
          loop(map)

      {:delete, key} ->
          loop(Map.delete(map, key))

    end
  end

  def concurrent_test(table, times) do
    Enum.each(1..times, fn it ->
      if Integer.is_even(it) do
        spawn fn ->
          send table, {:get, :counter, self()}
          v = receive do
            {:ok, r} -> r
          end
          send table, {:put, :counter, v+1}
        end
      else
        spawn fn ->
          send table, {:get, :counter, self()}
          v = receive do
           {:ok, r} -> r
          end
          send table, {:put, :counter, v-1}
        end
      end
    end)
  end

end


